### PR TITLE
CNS-Fix-Migrator

### PR DESCRIPTION
**Pull request by Yaroms**

### CODE REVIEW

File: x/dualstaking/keeper/hooks.go [hunk 1]
```diff
@@ -133,7 +133,7 @@ func (h Hooks) BeforeDelegationRemoved(ctx sdk.Context, delAddr sdk.AccAddress,
 	if err != nil {
 		return nil
 	}
-	amount := validator.TokensFromShares(delegation.Shares).TruncateInt()
+	amount := validator.TokensFromSharesRoundUp(delegation.Shares).TruncateInt()
 	err = h.k.UnbondUniformProviders(ctx, delAddr.String(), sdk.NewCoin(epochstoragetypes.TokenDenom, amount))
```
Comment (omerlavanet) [inline x/dualstaking/keeper/hooks.go:137]: if unbond uniform fails, do we really want to fail the hook? didn't we say we will allow some cases? maybe let's allow failing unbond uniform in the special case it's the last delegation, i think this should cover our edge case

File: x/dualstaking/keeper/migrations.go [hunk 2]
```diff
@@ -63,6 +62,12 @@ func (m Migrator) ConvertProviderStakeToSelfDelegation(ctx sdk.Context) error {
 		}
 	}
 
+	moduleBalance := m.keeper.bankKeeper.GetBalance(ctx, m.keeper.accountKeeper.GetModuleAddress(types.ModuleName), epochstoragetypes.TokenDenom)
+	err := m.keeper.bankKeeper.BurnCoins(ctx, types.ModuleName, sdk.NewCoins(moduleBalance))
```
Comment (omerlavanet) [inline x/dualstaking/keeper/migrations.go:66]: we are only burning dual staking, we will need to add pairing module burn in a migration (doesn't have to be in this one)

File: x/dualstaking/keeper/migrations.go [hunk 3]
```diff
@@ -74,14 +79,17 @@ func (m Migrator) HandleProviderDelegators(ctx sdk.Context) error {
...
-			return err
-		}
+	// burn all coins from the pools
+	moduleBalance := m.keeper.bankKeeper.GetBalance(ctx, m.keeper.accountKeeper.GetModuleAddress(dualstakingtypes.NotBondedPoolName), epochstoragetypes.TokenDenom)
+	err := m.keeper.bankKeeper.BurnCoins(ctx, dualstakingtypes.NotBondedPoolName, sdk.NewCoins(moduleBalance))
+	if err != nil {
+		return err
+	}
+
+	moduleBalance = m.keeper.bankKeeper.GetBalance(ctx, m.keeper.accountKeeper.GetModuleAddress(dualstakingtypes.NotBondedPoolName), epochstoragetypes.TokenDenom)
```
Comment (omerlavanet) [inline x/dualstaking/keeper/migrations.go:89]: why are we burning twice the same NotBondedPoolName looks like a mistake

**Review by omerlavanet — CHANGES_REQUESTED** [review]

File: x/dualstaking/keeper/hooks.go (hunk 1 above)
Comment (Yaroms) [reply x/dualstaking/keeper/hooks.go:137]: unbonduniform will fail only if something is wrong with the delegations,
so if we want to unbond more than we have it wont sent an error

File: x/dualstaking/keeper/migrations.go (hunk 2 above)
Comment (Yaroms) [reply x/dualstaking/keeper/migrations.go:66]: why?
mpdule pairing after this migration shouldnt hold any coins

File: x/dualstaking/keeper/migrations.go (hunk 3 above)
Comment (Yaroms) [reply x/dualstaking/keeper/migrations.go:89]: done

**Review by omerlavanet — APPROVED** [review]

**Review by omerlavanet — APPROVED** [review]

**Pull request merged**